File: aoc23/java/src/test/java/tech/dobler/aoc23/Day09Test.java
package tech.dobler.aoc23;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.stream.Stream;

import static org.assertj.core.api.Assertions.assertThat;

class Day09Test {

    final String testInput = """
            0 3 6 9 12 15
            1 3 6 10 15 21
            10 13 16 21 30 45
            """;
    private Day09 day;

    @BeforeEach
    void setUp() {
        day = new Day09();
    }

    private static Stream<String> readInput() {
        return Util.readLines(9);
    }

    @Test
    void part1() {
        assertThat(day.part1(Util.splitByNewLine(testInput))).isEqualTo(114L);

        assertThat(day.part1(readInput())).isEqualTo(2_098_530_125L);
    }

    @Test
    void part2() {
        assertThat(day.part2(Util.splitByNewLine(testInput))).isEqualTo(2L);

        assertThat(day.part2(readInput())).isEqualTo(1_016L);
    }
}

File: aoc21/Kotlin/src/test/kotlin/day09/Day09Test.kt
package day09

import org.junit.jupiter.api.Test
import util.readInput
import kotlin.test.assertEquals

class Day09Test {

    private val path = "09_test"

    @Test
    fun `parse input`()
    {
        val input = readInput(path, ::mapInput);
        assertEquals(
            listOf(
                listOf(2, 1, 9, 9, 9, 4, 3, 2, 1, 0),
                listOf(3, 9, 8, 7, 8, 9, 4, 9, 2, 1),
                listOf(9, 8, 5, 6, 7, 8, 9, 8, 9, 2),
                listOf(8, 7, 6, 7, 8, 9, 6, 7, 8, 9),
                listOf(9, 8, 9, 9, 9, 6, 5, 6, 7, 8)), input)
    }

    @Test
    fun `part 1 example`()
    {
        val input = readInput(path, ::mapInput);
        assertEquals(15, part1(input))
    }

    @Test
    fun `part 2 example`()
    {
        val input = readInput(path, ::mapInput);
        assertEquals(1134, part2(input))
    }
}